src/core/config.c

```diff
@@ -28,63 +30,57 @@ void finish_config(struct xdpw_config *config) {
 	free(config->screencast_conf.chooser_cmd);
 }
 
-static void getstring_from_conffile(dictionary *d,
-		const char *key, char **dest, const char *fallback) {
-	if (*dest != NULL) {
+static void parse_string(char **dest, const char* value) {
+	if (value == NULL || *value == '\0') {
+		logprint(TRACE, "config: skipping empty value in config file");
 		return;
 	}
-	const char *c = iniparser_getstring(d, key, fallback);
-	if (c == NULL) {
-		return;
-	}
-	// Allow keys without value as default
-	if (strcmp(c, "") != 0) {
-		*dest = strdup(c);
-	} else {
-		*dest = fallback ? strdup(fallback) : NULL;
-	}
+	free(*dest);
+	*dest = strdup(value);
 }
 
-static void getdouble_from_conffile(dictionary *d,
-		const char *key, double *dest, double fallback) {
-	if (*dest != 0) {
+static void parse_double(double *dest, const char* value) {
+	if (value == NULL || *value == '\0') {
+		logprint(TRACE, "config: skipping empty value in config file");
 		return;
 	}
-	*dest = iniparser_getdouble(d, key, fallback);
-}
-
-static bool file_exists(const char *path) {
-	return path && access(path, R_OK) != -1;
+	*dest = strtod(value, (char**)NULL);
 }
 
-static void config_parse_file(const char *configfile, struct xdpw_config *config) {
-	dictionary *d = NULL;
-	if (configfile) {
-		logprint(INFO, "config: using config file %s", configfile);
-		d = iniparser_load(configfile);
-	} else {
-		logprint(INFO, "config: no config file found");
-	}
-	if (configfile && !d) {
-		logprint(ERROR, "config: unable to load config file %s", configfile);
-	}
-
-	// screencast
-	getstring_from_conffile(d, "screencast:output_name", &config->screencast_conf.output_name, NULL);
-	getdouble_from_conffile(d, "screencast:max_fps", &config->screencast_conf.max_fps, 0);
-	getstring_from_conffile(d, "screencast:exec_before", &config->screencast_conf.exec_before, NULL);
-	getstring_from_conffile(d, "screencast:exec_after", &config->screencast_conf.exec_after, NULL);
-	getstring_from_conffile(d, "screencast:chooser_cmd", &config->screencast_conf.chooser_cmd, NULL);
-	if (!config->screencast_conf.chooser_type) {
+static int ini_parser_handler(void *data, const char* section, const char* name, const char* value) {
+	struct xdpw_config *config = (struct xdpw_config*)data;
+	logprint(TRACE, "config: parsing setction %s, key %s, value %s", section, name, value);
+
+	#define MATCH(s, n) strcmp(section, s) == 0 && strcmp(name, n) == 0
```

Comment:
This can be a function, bringing type safety in.

Reply:
If the `section` repetition is too verbose, we could have one handler per section:

```c
static int handle_ini_screencast(struct xdpw_config *config, const char *name, const char *value);
```

Reply:
About macros: https://drewdevault.com/2017/03/15/How-I-learned-to-stop-worrying-and-love-C.html